File: python3/copy_list_with_random_pointer.py
from .model import RandListNode

"""
Space   : O(2n)
Time    : O(2n)
"""


class Solution:
    def copyRandomList(self, head: RandListNode) -> RandListNode:
        p = head

        # create interweaved list
        while p:
            # save value to next creation
            value = p.val
            temp = p.next

            # insert copied current node
            p.next = RandListNode(value)
            p = p.next

            # point back to original next
            p.next = temp
            p = p.next

        # migrate random node
        p = head
        while p:
            if p.random:
                # save rand
                rand = p.random

                # insert rand to new rand
                p = p.next
                p.random = rand.next
                p = p.next
            else:
                p = p.next.next

        # create new node
        ans = RandListNode(0)
        a = ans
        p = head
        while p:
            # move next
            p = p.next

            # insert to ans
            a.next = p

            # move to next element
            a = a.next
            p = p.next

        return ans.next

File: golang/copy_list_with_random_pointer.go
package main

import (
	"golang/model"
)

// Space   : O(2n)
// Time    : O(2n)

func copyRandomList(head *model.RandListNode) *model.RandListNode {
	p := head

	for p != nil {
		// save value to next creation
		value := p.Val
		temp := p.Next

		// insert copied current node
		p.Next = &model.RandListNode{
			Val: value,
		}
		p = p.Next

		// point back to original next
		p.Next = temp
		p = p.Next
	}

	// migrate random node
	p = head
	for p != nil {
		if p.Random != nil {
			// save rand
			rand := p.Random

			// insert rand to new rand
			p = p.Next
			p.Random = rand.Next
			p = p.Next
		} else {
			p = p.Next.Next
		}
	}

	// create new node
	ans := &model.RandListNode{
		Val: 0,
	}
	pa := ans
	p = head
	for p != nil {
		// move next
		p = p.Next

		// insert to ans
		pa.Next = p

		// move to next element
		pa = pa.Next
		p = p.Next
	}

	return ans.Next
}
